raulin0/backup-juggler
# release: releases version 0.2.0 to the master branch

## Description

This pull request includes updates and additions for version 0.2.0 of the project. It introduces new features, documentation improvements, build configurations, and testing enhancements. The changes aim to improve the project's overall functionality, user experience and maintainability.

## Changes

- **Bump:** Version 0.1.0 → 0.2.0
- **docs:** Updates README.md for Backup Juggler
  - Added project overview and description
  - Included installation instructions
  - Provided links to documentation resources
  - Added information on how to contribute and ask for help

- **build:** Updates project metadata and configurations
  - Added license, classifiers, and URLs

- **docs:** Adds .readthedocs.yml configuration file
  - Added .readthedocs.yml configuration file to configure the documentation build process for Read the Docs. The file includes the necessary environment setup, installation of dependencies, and post-installation steps using Poetry.

- **ci:** Set up GitHub Actions pipeline for testing and code coverage
  - Added a GitHub Actions pipeline configuration file to the repository. The pipeline is triggered on push and pull_request events. It performs the following steps:
    - Copies the repository files
    - Installs Python 3.11.1
    - Installs the Poetry package manager
    - Installs project dependencies using Poetry
    - Runs tests and generates a coverage report in XML format
    - Uploads the coverage report to Codecov using the codecov/codecov-action

- **feat:** Implements the command line interface
- **docs:** Updates contribution guide for clarity and completeness
  - Introduced significant enhancements to the contribution guide to ensure it provides clear, comprehensive, and well-structured instructions for contributors. The changes include:
    - Added a high-level overview of the project structure
    - Detailed the purpose and functionality of each directory
    - Provided a comprehensive breakdown of the backup_juggler directory
    - Emphasized the use of the Typer library for building the CLI
    - Mentioned the use of the Rich library for generating detailed output
    - Described the testing setup, including the use of pytest and pytest-cov
    - Provided an overview of the documentation structure using mkdocs
    - Clarified the usage of the mkdocstrings plugin for generating API documentation
    - Added a "Tools" section to outline the main tools used
    - Outlined the contribution process step-by-step
    - Expanded on the Git Flow workflow
    - Advised contributors to perform code verification and run tests locally
    - Recommended the use of Commitizen for structured commit messages
    - Added instructions for pushing changes and opening a pull request
    - Included a reminder for contributors to seek assistance by opening an issue on the project's GitHub repository
    - Encouraged continuous improvement of the contribution guide

- **docs:** Updates MkDocs configuration
  - Updated the MkDocs configuration file with the following changes:
    - Added markdown extensions for attr_list, admonition, pymdownx.tasklist, and pymdownx.superfences
    - Included an extra CSS file
    - Added a custom run command in the extra section

- **docs:** Update index with Backup Juggler quick start information
  - Added comprehensive and detailed documentation for the Backup Juggler command-line tool

- **docs:** Adds initial version of tutorial, installation instructions, and styling
  - Combined the changes made to tutorial.md, installation.md, and extra.css into a single commit

- **build:** Updates project dependencies and add script entry
  - Updated dependencies and added a new script entry

- **feat:** Adds Backup Juggler logo
  - Added a logo to enhance the project's visual identity and recognition

## Motivation and Context

This pull request addresses several key areas to improve the project and align it with industry standards and best practices. The addition of new features, documentation enhancements, build configurations, and testing improvements contribute to the overall quality, usability, and maintainability of the project. The changes aim to provide a better user experience, attract contributors, and facilitate the future development and growth of the project.

## How to Test

1. Clone the repository and switch to the branch associated with this pull request.
2. Follow the updated installation instructions provided in the README.md file.
3. Familiarize yourself with the project structure and codebase by referring to the contribution guide in the documentation.
4. Explore the command-line interface by running various backup_juggler commands and options.
5. Verify that the documentation reflects the changes made in this pull request.
6. Run the test suite using the provided instructions and ensure that all tests pass.
7. Verify that the GitHub Actions pipeline for testing and code coverage executes successfully and generates the expected results.
8. Test the Backup Juggler CLI functionality thoroughly, including the newly added features and options.
9. Verify that the documentation is accurate, up-to-date, and covers all relevant aspects of the project.
10. Attempt different contribution scenarios following the updated contribution guide, from forking the repository to opening a pull request.
11. Seek assistance from the project maintainers or community members if needed.

## Related Issue(s)

N/A

## Checklist

- [x] All changes have been tested thoroughly.
- [x] Documentation has been updated to reflect the new features, configurations, and enhancements.
- [x] Code follows the project's style guidelines and conventions.
- [x] The GitHub Actions pipeline for testing and code coverage executes successfully.
- [x] The contribution guide provides clear and comprehensive instructions.
- [x] The installation instructions are accurate and up-to-date.
- [x] The Backup Juggler CLI is fully functional and has been tested extensively.

Please review and merge this pull request at your earliest convenience. Feel free to provide any feedback or suggestions for improvement. Thank you!
